### PR TITLE
Better requirements.txt handling

### PR DESCRIPTION
**Pull request by greysteil**
Previously, parsing and recreating a Python `requirements.txt` file only handled
the simple case of all dependencies being fully specified. This commit:
- Parses Python requirement files in the general case
- Adjusts the Python file update strategy to only alter a single line in the
  dependency file when regenerating it
- Ensures any comments and version specificity is preserved when a line in
  `requirements.txt` is updated

### CODE REVIEW

**Comment by greysteil** [comment]
@hmarr - all yours.
@hmac - FYI - will push this and a few others back upstream once the bigger refactor is merged

File: lib/bump/dependency_file_parsers/python.rb [hunk 1]
```diff
@@ -14,9 +13,60 @@ def initialize(dependency_files:)
...
+        attr_reader :line
+
+        def initialize(line)
+          @line = line
+        end
+
+        def parse
+          return if line.chomp.match(REQUIREMENT_LINE).nil?
+
+          dependency_name = Regexp.last_match[:name]
```
Comment (hmarr) [inline lib/bump/dependency_file_parsers/python.rb]: Not a huge fan of `Regexp.last_match` as it relies on nasty global state. Assign the return value from the call to `match` instead?

File: lib/bump/dependency_file_parsers/python.rb [hunk 2]
```diff
@@ -14,9 +13,60 @@ def initialize(dependency_files:)
...
-          parse_requirements(@requirements.content).
-          map { |dep| Dependency.new(name: dep[0], version: dep[1]) }
+        @requirements.
+          content.
+          each_line.
+          each_with_object([]) do |line, dependencies|
+            dependency = LineParser.new(line).parse
+
+            next if dependency.nil?
+            next if dependency[:requirements].count.zero?
```
Comment (hmarr) [inline lib/bump/dependency_file_parsers/python.rb]: I think `length` is generally preferred to `count` as it's faster.

File: lib/bump/dependency_file_parsers/python.rb [hunk 3]
```diff
@@ -14,9 +13,60 @@ def initialize(dependency_files:)
...
+
+            # Ignore dependencies with multiple requirements, since they would
+            # cause trouble at the gem update step
+            next if dependency[:requirements].count > 1
```
Comment (hmarr) [inline lib/bump/dependency_file_parsers/python.rb]: Same here

File: lib/bump/dependency_file_parsers/python.rb [hunk 4]
```diff
@@ -14,9 +13,60 @@ def initialize(dependency_files:)
...
+
+            version = dependency[:requirements].first.match(/[\d\.]+/)[0]
```
Comment (hmarr) [inline lib/bump/dependency_file_parsers/python.rb]: Any reason you're not using same regex as `LineParser::VERSION`?

File: lib/bump/dependency_file_updaters/python.rb [hunk 5]
```diff
@@ -32,20 +33,23 @@ def updated_requirements_file
...
-          )
-        end
-
-        @updated_requirements_content = packages.map do |pkg|
-          "#{pkg.name}==#{pkg.version}"
-        end.join("\n") + "\n"
+        requirements.content.
+          to_enum(:scan,
+                  DependencyFileParsers::Python::LineParser::REQUIREMENT_LINE).
+          find { Regexp.last_match[:name] == dependency.name }
```
Comment (hmarr) [inline lib/bump/dependency_file_updaters/python.rb]: I find this bit really hard to follow - it only works correctly because `to_enum` returns a _lazy_ enumerator, so the `Regexp.last_match` global state is correct at each iteration.

Also the fact that the result of the expression isn't assigned to anything then `$&` is used below isn't particularly clear.

Maybe just move to a simple call to `find` with a block that does the match, and return the result?

File: lib/bump/dependency_file_updaters/python.rb [hunk 6]
```diff
@@ -32,20 +33,23 @@ def updated_requirements_file
...
+
+        original_dep_declaration_string = $&
+        updated_dep_declaration_string =
+          original_dep_declaration_string.sub(/(,.*?)[\d\.]+/) do |old_version|
```
Comment (hmarr) [inline lib/bump/dependency_file_updaters/python.rb]: Again, will this regex suffice or do we need the one from `LineParser`?

File: lib/bump/dependency_file_updaters/python.rb [hunk 7]
```diff
@@ -32,20 +33,23 @@ def updated_requirements_file
...
+            matcher = Regexp.last_match[1]
```
Comment (hmarr) [inline lib/bump/dependency_file_updaters/python.rb]: Please no more `last_match`!

File: lib/bump/dependency_file_updaters/python.rb [hunk 8]
```diff
@@ -32,20 +33,23 @@ def updated_requirements_file
...
+            matcher + dependency.version.split(".").first(precision).join(".")
+          end
+
+        @updated_requirements_content = requirements.content.gsub(
+          original_dep_declaration_string,
+          updated_dep_declaration_string
+        )
       end
 
       def updated_version_string(old_version_string, new_version_number)
```
Comment (hmarr) [inline lib/bump/dependency_file_updaters/python.rb]: Should this be deleted now?
